### PR TITLE
RELATED: RAIL-3699, RAIL-3700, RAIL-3705 insight and dashboard user info

### PR DESCRIPTION
**Pull request by no23reason**
Adds the possibility for user information to be included when retrieving insights and dashboards (introducting a new universal `IAuditable` interface).
Also adds `all` and `allSorted` methods to all `PagedResource` instances that allow getting all the data of a given paged resource. This is useful for custom client side sorting logic (something that the backend does not support).
Finally, makes the `getInsights` function on bear faster by saving redundant visClasses call for every subsequent page there.

---

Supported PR commands:

| Command                  | Description            |
| ------------------------ | ---------------------- |
| `ok to test`             | Re-run standard checks |
| `extended test`          | BackstopJS tests       |
| `extended check sonar`   | SonarQube tests        |
| `extended check cypress` | Cypress E2E tests      |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5186/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6893/ : FAILURE in 5m 27s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5187/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6894/ : FAILURE in 14m 42s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5188/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6895/ : SUCCESS in 14m 46s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5189/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6896/ : SUCCESS in 14m 35s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5193/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6899/ : FAILURE in 14m 48s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5194/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6900/ : SUCCESS in 15m 08s

File: libs/sdk-backend-base/src/toolkit/paging.ts [hunk 1]
```diff
@@ -14,34 +14,84 @@ export class InMemoryPaging<T> implements IPagedResource<T> {
...
         if (this.items.length === 0) {
             return this;
         }
 
-        return new InMemoryPaging(this.all, this.limit, pageIndex * this.items.length);
+        return new InMemoryPaging(this.allItems, this.limit, pageIndex * this.items.length);
     }
+
+    public async all(): Promise<T[]> {
+        return this.allItems;
```
Comment (lupko) [inline libs/sdk-backend-base/src/toolkit/paging.ts:49]: please return copy of array here

File: libs/sdk-backend-base/src/toolkit/paging.ts [hunk 2]
```diff
@@ -14,34 +14,84 @@ export class InMemoryPaging<T> implements IPagedResource<T> {
...
+    }
+
+    public async allSorted(compareFn: (a: T, b: T) => number): Promise<T[]> {
+        return this.allItems.sort(compareFn);
```
Comment (lupko) [inline libs/sdk-backend-base/src/toolkit/paging.ts:53]: same here

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: libs/sdk-backend-base/src/toolkit/paging.ts (hunk 1 above)
Comment (no23reason) [reply libs/sdk-backend-base/src/toolkit/paging.ts]: Fixed in dce269b

File: libs/sdk-backend-base/src/toolkit/paging.ts (hunk 2 above)
Comment (no23reason) [reply libs/sdk-backend-base/src/toolkit/paging.ts]: Fixed in dce269b

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/5198/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6904/ : SUCCESS in 16m 55s

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1744

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/1483/ : SUCCESS in 53s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1760/ : SUCCESS in 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/547/ : SUCCESS in 18m 12s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6911/ : SUCCESS in 14m 44s

**Pull request merged**